File: napkin/src/patchpanel/plugitem.cpp
#include "plugitem.h"
#include <QGraphicsScene>

#include "../napkin_utilities.h"

PlugItem::PlugItem(OperatorItem* operatorItem, nap::Plug& plug)
	: QGraphicsItem(operatorItem), mOperator(operatorItem), mConnectedWire(NULL), mPlug(plug)
{
	mLabel = new QGraphicsTextItem(this);
	mPin = new PinItem(this);
	setParentItem(operatorItem);

	setName(QString::fromStdString(plug.getName()));
}

QRectF PlugItem::boundingRect() const { return childrenBoundingRect(); }

void PlugItem::setName(const QString name)
{
	mLabel->setPlainText(name);
	layout();
}
void PlugItem::layout()
{
	double textoffset = -6;

	if (isInput()) {
		mPin->setPos(0, 0);
		mLabel->setPos(mPin->boundingRect().width(), textoffset);
		//		mValue->setPos(-mValue->boundingRect().width(), 0);
	} else {
		mLabel->setPos(0, textoffset);
		mPin->setPos(mLabel->boundingRect().width(), 0);
	}

	mOperator->layout();
}

File: py/patch/plugitem.py
from PyQt5.QtWidgets import QGraphicsItem, QGraphicsTextItem

import nap
from patch.pinitem import PinItem


class PlugItem(QGraphicsItem):
    def __init__(self, opItem, plug):
        super(PlugItem, self).__init__()
        self.__plug = plug
        self.__opItem = opItem
        self.__label = QGraphicsTextItem(self)
        self.__pin = PinItem(self)
        self.setParentItem(self.__opItem)
        self.setName(plug.name())

    def setName(self, name):
        self.__label.setPlainText(name)
        self.layout()

    def boundingRect(self):
        return self.childrenBoundingRect()

    def paint(self, painter, option, widget=None):
        pass

    def pin(self):
        return self.__pin

    def plug(self):
        """
        @rtype: nap.Plug
        """
        return self.__plug

    def isInput(self):
        return isinstance(self.__plug, nap.InputPlugBase)

    def setUsable(self, b):
        self.setVisible(b)

    def layout(self):
        # textOffset = 6

        pinY = self.__label.boundingRect().height() / 2 - self.__pin.boundingRect().height() / 2

        if self.isInput():
            self.__label.setPos(self.pin().boundingRect().width(), 0)
            self.__pin.setPos(0, pinY)
        else:
            self.__label.setPos(0, 0)
            self.__pin.setPos(self.__label.boundingRect().width(), pinY)

        self.__opItem.layout()